File: src/lib/browser/human-behavior.ts
/**
 * Advanced Human Behavior Simulation
 * 
 * Simulates realistic human interaction patterns to avoid bot detection.
 * 
 * Features:
 * - Bezier curve mouse movements (mimics real hand movement)
 * - Variable typing speeds with realistic typos
 * - Natural scrolling patterns
 * - Random pauses that mimic reading/thinking
 * - Focus/blur simulation
 * - Realistic click patterns
 */

import type { Page } from 'playwright';

// ===========================================
// Types
// ===========================================

export interface HumanBehaviorConfig {
  // Typing
  typingSpeed: 'slow' | 'normal' | 'fast' | 'variable';
  baseTypingDelayMs: number;
  typingVariance: number;
  typoRate: number; // Probability of making a typo
  
  // Mouse
  mouseMovementStyle: 'bezier' | 'linear' | 'natural';
  mouseSpeed: 'slow' | 'normal' | 'fast';
  jitterAmount: number;
  
  // Scrolling
  scrollStyle: 'smooth' | 'stepped' | 'variable';
  scrollSpeed: 'slow' | 'normal' | 'fast';
  
  // Pauses
  readingPauseMs: { min: number; max: number };
  thinkingPauseMs: { min: number; max: number };
  
  // Clicks
  clickHoldMs: { min: number; max: number };
  doubleClickProbability: number;
}

// ===========================================
// Default Configuration
// ===========================================

export const DEFAULT_HUMAN_CONFIG: HumanBehaviorConfig = {
  typingSpeed: 'variable',
  baseTypingDelayMs: 80,
  typingVariance: 0.5,
  typoRate: 0.02,
  
  mouseMovementStyle: 'bezier',
  mouseSpeed: 'normal',
  jitterAmount: 2,
  
  scrollStyle: 'variable',
  scrollSpeed: 'normal',
  
  readingPauseMs: { min: 1000, max: 3000 },
  thinkingPauseMs: { min: 500, max: 1500 },
  
  clickHoldMs: { min: 50, max: 150 },
  doubleClickProbability: 0.02,
};

// ===========================================
// Human Behavior Simulator
// ===========================================

export class HumanBehavior {
  private config: HumanBehaviorConfig;
  private lastMousePosition: { x: number; y: number } = { x: 0, y: 0 };
  
  constructor(config: Partial<HumanBehaviorConfig> = {}) {
    this.config = { ...DEFAULT_HUMAN_CONFIG, ...config };
  }
  
  // ===========================================
  // Mouse Movements
  // ===========================================
  
  /**
   * Move mouse to element using bezier curve
   */
  async moveToElement(page: Page, selector: string): Promise<void> {
    const element = page.locator(selector);
    const box = await element.boundingBox();
    
    if (!box) {
      console.warn(`[HumanBehavior] Element not found: ${selector}`);
      return;
    }
    
    // Target with slight randomness (don't always hit center)
    const targetX = box.x + box.width * (0.3 + Math.random() * 0.4);
    const targetY = box.y + box.height * (0.3 + Math.random() * 0.4);
    
    await this.moveTo(page, targetX, targetY);
  }
  
  /**
   * Move mouse to coordinates using bezier curve
   */
  async moveTo(page: Page, targetX: number, targetY: number): Promise<void> {
    const startX = this.lastMousePosition.x || Math.random() * 500 + 100;
    const startY = this.lastMousePosition.y || Math.random() * 300 + 100;
    
    // Generate bezier control points
    const distance = Math.sqrt(Math.pow(targetX - startX, 2) + Math.pow(targetY - startY, 2));
    const duration = this.calculateMoveDuration(distance);
    const steps = Math.max(20, Math.floor(distance / 10));
    
    // Create bezier curve with 2 control points
    const cp1 = {
      x: startX + (targetX - startX) * 0.3 + (Math.random() - 0.5) * distance * 0.2,
      y: startY + (targetY - startY) * 0.3 + (Math.random() - 0.5) * distance * 0.2,
    };
    const cp2 = {
      x: startX + (targetX - startX) * 0.7 + (Math.random() - 0.5) * distance * 0.2,
      y: startY + (targetY - startY) * 0.7 + (Math.random() - 0.5) * distance * 0.2,
    };
    
    // Move along bezier curve
    for (let i = 0; i <= steps; i++) {
      const t = i / steps;
      
      // Cubic bezier formula
      const x = this.cubicBezier(t, startX, cp1.x, cp2.x, targetX);
      const y = this.cubicBezier(t, startY, cp1.y, cp2.y, targetY);
      
      // Add micro-jitter
      const jitterX = (Math.random() - 0.5) * this.config.jitterAmount;
      const jitterY = (Math.random() - 0.5) * this.config.jitterAmount;
      
      await page.mouse.move(x + jitterX, y + jitterY);
      
      // Variable delay (faster in middle, slower at ends)
      const speedMultiplier = 1 - Math.abs(t - 0.5) * 0.5;
      const stepDelay = (duration / steps) * speedMultiplier;
      await this.sleep(stepDelay);
    }
    
    this.lastMousePosition = { x: targetX, y: targetY };
  }
  
  /**
   * Cubic bezier interpolation
   */
  private cubicBezier(t: number, p0: number, p1: number, p2: number, p3: number): number {
    const mt = 1 - t;
    return mt * mt * mt * p0 + 3 * mt * mt * t * p1 + 3 * mt * t * t * p2 + t * t * t * p3;
  }
  
  /**
   * Calculate duration for mouse movement based on distance
   */
  private calculateMoveDuration(distance: number): number {
    const baseSpeed = this.config.mouseSpeed === 'slow' ? 0.3 
                    : this.config.mouseSpeed === 'fast' ? 1.5 
                    : 0.7;
    
    // Fitts's Law inspired calculation
    const duration = (distance * 0.5 + 100) / baseSpeed;
    
    // Add randomness
    return duration * (0.8 + Math.random() * 0.4);
  }
  
  // ===========================================
  // Clicking
  // ===========================================
  
  /**
   * Click with human-like timing
   */
  async click(page: Page, selector: string): Promise<void> {
    // Move to element first
    await this.moveToElement(page, selector);
    
    // Small pause before clicking (aiming)
    await this.sleep(this.randomBetween(50, 150));
    
    // Click with hold time
    const holdTime = this.randomBetween(
      this.config.clickHoldMs.min,
      this.config.clickHoldMs.max
    );
    
    await page.mouse.down();
    await this.sleep(holdTime);
    await page.mouse.up();
    
    // Occasional double-click
    if (Math.random() < this.config.doubleClickProbability) {
      await this.sleep(this.randomBetween(80, 150));
      await page.mouse.down();
      await this.sleep(holdTime * 0.7);
      await page.mouse.up();
    }
    
    // Small pause after clicking
    await this.sleep(this.randomBetween(100, 300));
  }
  
  // ===========================================
  // Typing
  // ===========================================
  
  /**
   * Type text with human-like patterns
   */
  async type(page: Page, selector: string, text: string): Promise<void> {
    const element = page.locator(selector);
    await element.click();
    
    // Small pause before typing
    await this.sleep(this.randomBetween(200, 500));
    
    const chars = text.split('');
    let i = 0;
    
    while (i < chars.length) {
      const char = chars[i];
      
      // Simulate typo
      if (Math.random() < this.config.typoRate && i > 0) {
        const wrongChar = this.getAdjacentKey(char);
        await page.keyboard.type(wrongChar);
        await this.getTypingDelay();
        
        // Notice and correct the typo
        await this.sleep(this.randomBetween(200, 600)); // Pause to "notice"
        await page.keyboard.press('Backspace');
        await this.sleep(this.randomBetween(50, 150));
      }
      
      // Type the character
      await page.keyboard.type(char);
      
      // Variable delay
      let delay = this.getTypingDelay();
      
      // Longer pause after punctuation
      if (['.', '!', '?', ',', ';', ':'].includes(char)) {
        delay += this.randomBetween(100, 300);
      }
      
      // Pause at word boundaries (thinking)
      if (char === ' ' && Math.random() < 0.1) {
        delay += this.randomBetween(200, 800);
      }
      
      await this.sleep(delay);
      i++;
    }
  }
  
  /**
   * Get adjacent key for typo simulation
   */
  private getAdjacentKey(char: string): string {
    const keyboard: Record<string, string[]> = {
      'a': ['s', 'q', 'z'],
      'b': ['v', 'g', 'n'],
      'c': ['x', 'd', 'v'],
      'd': ['s', 'e', 'f', 'c'],
      'e': ['w', 'r', 'd'],
      'f': ['d', 'r', 'g', 'v'],
      'g': ['f', 't', 'h', 'b'],
      'h': ['g', 'y', 'j', 'n'],
      'i': ['u', 'o', 'k'],
      'j': ['h', 'u', 'k', 'm'],
      'k': ['j', 'i', 'l', 'm'],
      'l': ['k', 'o', 'p'],
      'm': ['n', 'j', 'k'],
      'n': ['b', 'h', 'm'],
      'o': ['i', 'p', 'l'],
      'p': ['o', 'l'],
      'q': ['w', 'a'],
      'r': ['e', 't', 'f'],
      's': ['a', 'w', 'd', 'z'],
      't': ['r', 'y', 'g'],
      'u': ['y', 'i', 'j'],
      'v': ['c', 'f', 'b'],
      'w': ['q', 'e', 's'],
      'x': ['z', 's', 'c'],
      'y': ['t', 'u', 'h'],
      'z': ['a', 'x'],
    };
    
    const lowerChar = char.toLowerCase();
    const adjacent = keyboard[lowerChar] || [char];
    const randomAdjacent = adjacent[Math.floor(Math.random() * adjacent.length)];
    
    // Maintain case
    return char === char.toUpperCase() ? randomAdjacent.toUpperCase() : randomAdjacent;
  }
  
  /**
   * Get variable typing delay
   */
  private getTypingDelay(): number {
    const base = this.config.baseTypingDelayMs;
    const variance = this.config.typingVariance;
    
    // Speed variations
    const speedMultiplier = this.config.typingSpeed === 'slow' ? 1.5
                         : this.config.typingSpeed === 'fast' ? 0.6
                         : this.config.typingSpeed === 'variable' ? (0.5 + Math.random())
                         : 1;
    
    // Calculate delay with variance
    const delay = base * speedMultiplier * (1 + (Math.random() - 0.5) * variance);
    
    return Math.max(30, delay);
  }
  
  // ===========================================
  // Scrolling
  // ===========================================
  
  /**
   * Scroll with human-like patterns
   */
  async scroll(page: Page, direction: 'up' | 'down', amount: number): Promise<void> {
    const actualAmount = direction === 'down' ? amount : -amount;
    
    if (this.config.scrollStyle === 'smooth') {
      await this.smoothScroll(page, actualAmount);
    } else if (this.config.scrollStyle === 'stepped') {
      await this.steppedScroll(page, actualAmount);
    } else {
      // Variable - mix of both
      if (Math.random() > 0.5) {
        await this.smoothScroll(page, actualAmount);
      } else {
        await this.steppedScroll(page, actualAmount);
      }
    }
  }
  
  /**
   * Smooth momentum-based scroll
   */
  private async smoothScroll(page: Page, amount: number): Promise<void> {
    const steps = Math.abs(amount) > 500 ? 30 : 15;
    const direction = amount > 0 ? 1 : -1;
    const totalAmount = Math.abs(amount);
    
    // Simulate scroll momentum (fast start, slow end)
    for (let i = 0; i < steps; i++) {
      const progress = i / steps;
      
      // Ease-out curve
      const easing = 1 - Math.pow(1 - progress, 3);
      const previousEasing = i > 0 ? 1 - Math.pow(1 - (i - 1) / steps, 3) : 0;
      
      const scrollAmount = (easing - previousEasing) * totalAmount * direction;
      
      await page.mouse.wheel(0, scrollAmount);
      
      // Variable delay
      const delay = 10 + Math.random() * 20 + (progress > 0.7 ? 20 : 0);
      await this.sleep(delay);
    }
    
    // Pause after scrolling (reading)
    await this.sleep(this.randomBetween(300, 800));
  }
  
  /**
   * Stepped scroll (discrete wheel ticks)
   */
  private async steppedScroll(page: Page, amount: number): Promise<void> {
    const tickAmount = 100; // Standard wheel tick
    const ticks = Math.ceil(Math.abs(amount) / tickAmount);
    const direction = amount > 0 ? 1 : -1;
    
    for (let i = 0; i < ticks; i++) {
      const scrollAmount = tickAmount * direction * (0.8 + Math.random() * 0.4);
      
      await page.mouse.wheel(0, scrollAmount);
      await this.sleep(this.randomBetween(50, 150));
    }
    
    // Pause after scrolling
    await this.sleep(this.randomBetween(500, 1500));
  }
  
  // ===========================================
  // Pauses & Waits
  // ===========================================
  
  /**
   * Pause as if reading content
   */
  async readingPause(page: Page): Promise<void> {
    const duration = this.randomBetween(
      this.config.readingPauseMs.min,
      this.config.readingPauseMs.max
    );
    
    // Maybe scroll a little while reading
    if (Math.random() > 0.6) {
      await this.sleep(duration * 0.3);
      await this.scroll(page, 'down', this.randomBetween(50, 150));
      await this.sleep(duration * 0.7);
    } else {
      await this.sleep(duration);
    }
  }
  
  /**
   * Pause as if thinking
   */
  async thinkingPause(page: Page): Promise<void> {
    const duration = this.randomBetween(
      this.config.thinkingPauseMs.min,
      this.config.thinkingPauseMs.max
    );
    
    // Maybe move mouse randomly while thinking
    if (Math.random() > 0.7) {
      const viewport = page.viewportSize() || { width: 1920, height: 1080 };
      const randomX = this.lastMousePosition.x + (Math.random() - 0.5) * 100;
      const randomY = this.lastMousePosition.y + (Math.random() - 0.5) * 50;
      
      await this.moveTo(
        page,
        Math.max(0, Math.min(viewport.width, randomX)),
        Math.max(0, Math.min(viewport.height, randomY))
      );
    }
    
    await this.sleep(duration);
  }
  
  /**
   * Wait with variance
   */
  async wait(baseMs: number, variance: number = 0.3): Promise<void> {
    const actual = baseMs * (1 + (Math.random() - 0.5) * variance * 2);
    await this.sleep(actual);
  }
  
  // ===========================================
  // Focus/Blur Simulation
  // ===========================================
  
  /**
   * Simulate user looking away (blur) and coming back (focus)
   */
  async simulateDistraction(page: Page): Promise<void> {
    // Blur
    await page.evaluate(() => {
      window.dispatchEvent(new Event('blur'));
      document.dispatchEvent(new Event('visibilitychange'));
    });
    
    // Wait (user is distracted)
    await this.sleep(this.randomBetween(2000, 8000));
    
    // Focus
    await page.evaluate(() => {
      window.dispatchEvent(new Event('focus'));
      document.dispatchEvent(new Event('visibilitychange'));
    });
    
    // Small pause to "reorient"
    await this.sleep(this.randomBetween(300, 800));
  }
  
  // ===========================================
  // Utilities
  // ===========================================
  
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
  
  private randomBetween(min: number, max: number): number {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }
}

// ===========================================
// Singleton
// ===========================================

let humanBehaviorInstance: HumanBehavior | null = null;

export function getHumanBehavior(config?: Partial<HumanBehaviorConfig>): HumanBehavior {
  if (!humanBehaviorInstance) {
    humanBehaviorInstance = new HumanBehavior(config);
  }
  return humanBehaviorInstance;
}

export default HumanBehavior;

File: rpa/utils/human_behavior.py
"""
Human-like behavior simulation for RPA automation.

This module provides functions to simulate human-like interactions:
- Variable typing speed
- Random mouse movements
- Natural pauses and delays
- Scroll patterns
"""

import random
import time
import math
from dataclasses import dataclass
from typing import Tuple, Optional
from playwright.sync_api import Page


@dataclass
class HumanBehaviorConfig:
    """Configuration for human-like behavior."""
    
    # Typing - varies per "person" for consistency within session
    base_typing_delay_ms: float = 80  # Base delay between keystrokes
    typing_variance: float = 0.4  # 40% variance
    word_pause_chance: float = 0.1  # 10% chance to pause at word boundaries
    word_pause_range: Tuple[float, float] = (100, 400)  # ms
    typo_chance: float = 0.0  # Set to 0 - we don't want actual typos
    
    # Mouse - more natural movement
    mouse_speed_factor: float = 1.0  # 1.0 = normal speed
    mouse_curve_steps: int = 25  # More points for smoother curve
    mouse_overshoot_chance: float = 0.15  # 15% chance to slightly overshoot
    
    # Scrolling - irregular patterns
    scroll_step_range: Tuple[int, int] = (40, 180)  # pixels per scroll
    scroll_delay_range: Tuple[float, float] = (30, 200)  # ms between scrolls
    scroll_pause_chance: float = 0.2  # 20% chance to pause mid-scroll
    
    # General pauses - more variety
    micro_pause_range: Tuple[float, float] = (50, 250)  # ms
    think_pause_range: Tuple[float, float] = (400, 2000)  # ms
    hesitation_chance: float = 0.1  # 10% chance of random hesitation
    
    # Fatigue simulation (typing slows down over time)
    enable_fatigue: bool = True
    fatigue_factor: float = 0.001  # Increase delay by 0.1% per character


class HumanBehavior:
    """
    Provides human-like behavior simulation for browser automation.
    
    Implements sophisticated patterns that mimic real human behavior:
    - Gaussian-distributed typing delays
    - Mouse movements with bezier curves and occasional overshoots
    - Reading/thinking pauses
    - Fatigue simulation (slowing down over time)
    - Natural hesitation and micro-corrections
    
    Usage:
        human = HumanBehavior(page)
        human.type_like_human(selector, "Hello world")
        human.click_like_human(selector)
    """
    
    def __init__(self, page: Page, config: Optional[HumanBehaviorConfig] = None):
        self.page = page
        self.config = config or HumanBehaviorConfig()
        self._chars_typed = 0  # Track for fatigue simulation
        self._last_mouse_pos: Optional[Tuple[float, float]] = None
    
    def _random_delay(self, min_ms: float, max_ms: float) -> float:
        """Generate a random delay in milliseconds."""
        return random.uniform(min_ms, max_ms)
    
    def _gaussian_delay(self, mean_ms: float, variance: float) -> float:
        """Generate a gaussian-distributed delay."""
        std_dev = mean_ms * variance
        delay = random.gauss(mean_ms, std_dev)
        return max(10, delay)  # Minimum 10ms
    
    def _get_fatigue_factor(self) -> float:
        """Calculate fatigue multiplier based on characters typed."""
        if not self.config.enable_fatigue:
            return 1.0
        return 1.0 + (self._chars_typed * self.config.fatigue_factor)
    
    def micro_pause(self) -> None:
        """A tiny pause, like natural hesitation."""
        delay = self._random_delay(*self.config.micro_pause_range)
        time.sleep(delay / 1000)
    
    def think_pause(self) -> None:
        """A longer pause, like the user is thinking."""
        delay = self._random_delay(*self.config.think_pause_range)
        time.sleep(delay / 1000)
    
    def hesitate(self) -> None:
        """Random hesitation - called occasionally during actions."""
        if random.random() < self.config.hesitation_chance:
            pause = random.uniform(0.2, 0.8)
            time.sleep(pause)
    
    def wait(self, min_sec: float, max_sec: float) -> None:
        """Wait for a random duration between min and max seconds."""
        delay = random.uniform(min_sec, max_sec)
        time.sleep(delay)
    
    def type_like_human(
        self,
        selector: str,
        text: str,
        click_first: bool = True
    ) -> None:
        """
        Type text with human-like delays and patterns.
        
        Features:
        - Gaussian-distributed inter-key delays
        - Longer pauses at word boundaries
        - Fatigue simulation (typing slows over time)
        - Occasional hesitations
        
        Args:
            selector: CSS selector for the input element
            text: Text to type
            click_first: Whether to click the element before typing
        """
        if click_first:
            self.click_like_human(selector)
            self.micro_pause()
        
        element = self.page.locator(selector).first
        
        for i, char in enumerate(text):
            # Calculate base delay with fatigue
            base_delay = self._gaussian_delay(
                self.config.base_typing_delay_ms,
                self.config.typing_variance
            ) * self._get_fatigue_factor()
            
            delay = base_delay
            
            # Add pause at word boundaries
            if char == " " and random.random() < self.config.word_pause_chance:
                extra_pause = self._random_delay(*self.config.word_pause_range)
                delay += extra_pause
            
            # Occasional hesitation mid-typing
            if random.random() < 0.02:  # 2% chance
                hesitation = random.uniform(100, 400)
                delay += hesitation
            
            # Slightly longer delay after punctuation (like thinking about next word)
            if i > 0 and text[i-1] in '.!?,;:':
                delay += random.uniform(100, 300)
            
            # Type the character
            element.type(char, delay=int(delay))
            self._chars_typed += 1
    
    def click_like_human(self, selector: str) -> None:
        """
        Click an element with human-like behavior.
        
        Features:
        - Random click position within element bounds
        - Natural mouse movement with bezier curves
        - Occasional overshoot and correction
        - Pre-click hesitation
        """
        element = self.page.locator(selector).first
        
        # Occasional hesitation before clicking
        self.hesitate()
        
        # Get element bounding box
        box = element.bounding_box()
        if box:
            # Random position within element (avoiding edges, weighted toward center)
            # Use gaussian distribution for more natural click positions
            margin = min(box["width"], box["height"]) * 0.15
            center_x = box["x"] + box["width"] / 2
            center_y = box["y"] + box["height"] / 2
            
            # Gaussian distribution centered on element center
            x = random.gauss(center_x, box["width"] * 0.15)
            y = random.gauss(center_y, box["height"] * 0.15)
            
            # Clamp to element bounds with margin
            x = max(box["x"] + margin, min(box["x"] + box["width"] - margin, x))
            y = max(box["y"] + margin, min(box["y"] + box["height"] - margin, y))
            
            # Move mouse with curve (possibly with overshoot)
            self._move_mouse_with_curve(x, y, may_overshoot=True)
            
            # Micro pause before clicking (humans don't click instantly)
            time.sleep(random.uniform(0.05, 0.15))
            
            # Click
            self.page.mouse.click(x, y)
            
            # Track position
            self._last_mouse_pos = (x, y)
        else:
            # Fallback to regular click
            element.click()
    
    def _move_mouse_with_curve(
        self, 
        target_x: float, 
        target_y: float,
        may_overshoot: bool = False
    ) -> None:
        """
        Move mouse to target with a natural bezier curve.
        
        Features:
        - Smooth bezier curve movement
        - Variable speed (slower at start/end)
        - Occasional overshoot and correction
        - Jitter for more natural appearance
        """
        # Get starting position
        if self._last_mouse_pos:
            start_x, start_y = self._last_mouse_pos
            # Add small random offset to start (hand tremor)
            start_x += random.uniform(-3, 3)
            start_y += random.uniform(-3, 3)
        else:
            # Approximate from viewport
            viewport = self.page.viewport_size or {"width": 1280, "height": 720}
            start_x = viewport["width"] / 2 + random.uniform(-100, 100)
            start_y = viewport["height"] / 2 + random.uniform(-100, 100)
        
        # Calculate distance for speed adjustment
        distance = math.sqrt((target_x - start_x)**2 + (target_y - start_y)**2)
        
        # Adjust steps based on distance (more steps for longer movements)
        base_steps = self.config.mouse_curve_steps
        steps = max(10, int(base_steps * (distance / 300)))
        
        # Generate bezier control points with natural curves
        # Control points create a slightly curved path, not straight line
        angle_offset = random.uniform(-0.3, 0.3)  # Slight curve to one side
        
        control_x1 = start_x + (target_x - start_x) * 0.25 + random.uniform(-40, 40)
        control_y1 = start_y + (target_y - start_y) * 0.1 + random.uniform(-25, 25)
        control_x2 = start_x + (target_x - start_x) * 0.75 + random.uniform(-25, 25)
        control_y2 = start_y + (target_y - start_y) * 0.9 + random.uniform(-15, 15)
        
        # Add curve bias
        perpendicular_x = -(target_y - start_y) * angle_offset * 0.3
        perpendicular_y = (target_x - start_x) * angle_offset * 0.3
        control_x1 += perpendicular_x
        control_y1 += perpendicular_y
        control_x2 += perpendicular_x * 0.5
        control_y2 += perpendicular_y * 0.5
        
        # Move along curve
        for i in range(steps + 1):
            t = i / steps
            
            # Cubic bezier formula
            x = (
                (1 - t) ** 3 * start_x +
                3 * (1 - t) ** 2 * t * control_x1 +
                3 * (1 - t) * t ** 2 * control_x2 +
                t ** 3 * target_x
            )
            y = (
                (1 - t) ** 3 * start_y +
                3 * (1 - t) ** 2 * t * control_y1 +
                3 * (1 - t) * t ** 2 * control_y2 +
                t ** 3 * target_y
            )
            
            # Add micro-jitter (hand tremor) - more prominent at slow parts
            if random.random() < 0.3:
                jitter = 1.5 * (1 - abs(2 * t - 1))  # More jitter in middle
                x += random.uniform(-jitter, jitter)
                y += random.uniform(-jitter, jitter)
            
            self.page.mouse.move(x, y)
            
            # Variable speed - slower at start and end (acceleration curve)
            # Using sine wave for smooth acceleration/deceleration
            speed_factor = 0.5 + 0.5 * math.sin(math.pi * t)
            speed_factor = max(0.3, speed_factor)  # Don't go too slow
            
            base_delay = 4 / self.config.mouse_speed_factor
            delay = base_delay / speed_factor
            
            # Add occasional micro-pause (hand repositioning)
            if random.random() < 0.05:
                delay += random.uniform(20, 50)
            
            time.sleep(delay / 1000)
        
        # Overshoot and correct (happens naturally with humans)
        if may_overshoot and random.random() < self.config.mouse_overshoot_chance:
            # Overshoot by a small amount
            overshoot_x = target_x + random.uniform(5, 15) * (1 if random.random() > 0.5 else -1)
            overshoot_y = target_y + random.uniform(5, 15) * (1 if random.random() > 0.5 else -1)
            
            self.page.mouse.move(overshoot_x, overshoot_y)
            time.sleep(random.uniform(0.05, 0.12))
            
            # Correct back to target
            self.page.mouse.move(target_x, target_y)
            time.sleep(random.uniform(0.02, 0.05))
        
        self._last_mouse_pos = (target_x, target_y)
    
    def scroll_down(self, amount: int = 300) -> None:
        """
        Scroll down with human-like behavior.
        
        Args:
            amount: Approximate amount to scroll in pixels
        """
        scrolled = 0
        while scrolled < amount:
            step = random.randint(*self.config.scroll_step_range)
            step = min(step, amount - scrolled)
            
            self.page.mouse.wheel(0, step)
            scrolled += step
            
            delay = self._random_delay(*self.config.scroll_delay_range)
            time.sleep(delay / 1000)
    
    def scroll_up(self, amount: int = 300) -> None:
        """Scroll up with human-like behavior."""
        scrolled = 0
        while scrolled < amount:
            step = random.randint(*self.config.scroll_step_range)
            step = min(step, amount - scrolled)
            
            self.page.mouse.wheel(0, -step)
            scrolled += step
            
            delay = self._random_delay(*self.config.scroll_delay_range)
            time.sleep(delay / 1000)
    
    def scroll_into_view(self, selector: str) -> None:
        """Scroll an element into view naturally."""
        element = self.page.locator(selector).first
        
        # Check if element is in viewport
        box = element.bounding_box()
        viewport = self.page.viewport_size or {"width": 1280, "height": 720}
        
        if box:
            if box["y"] < 0:
                # Element is above viewport
                self.scroll_up(int(abs(box["y"]) + 100))
            elif box["y"] + box["height"] > viewport["height"]:
                # Element is below viewport
                self.scroll_down(int(box["y"] - viewport["height"] / 2))

